core/math/geometry_3d.h

```diff
@@ -532,6 +533,19 @@ class Geometry3D {
 		return clipped;
 	}
 
+	static Vector<int32_t> tetrahedralize_delaunay(const Vector<Vector3> &p_points) {
+		Vector<Delaunay3D::OutputSimplex> tetr = Delaunay3D::tetrahedralize(p_points);
+		Vector<int32_t> tetrahedrons;
+
+		for (int i = 0; i < tetr.size(); i++) {
+			tetrahedrons.push_back(tetr[i].points[0]);
+			tetrahedrons.push_back(tetr[i].points[1]);
+			tetrahedrons.push_back(tetr[i].points[2]);
+			tetrahedrons.push_back(tetr[i].points[3]);
+		}
```

Comment:
```suggestion
		tetrahedrons.resize(4 * tetr.size());
		int32_t *ptr = tetrahedrons.ptrw();
		for (int i = 0; i < tetr.size(); i++) {
			*ptr++ = tetr[i].points[0];
			*ptr++ = tetr[i].points[1];
			*ptr++ = tetr[i].points[2];
			*ptr++ = tetr[i].points[3];
		}
```
Avoids reallocation

Reply:
Before accepting can I ask what is the `ptrw()` method for? I tried searching for any definition but to no avail (given the name I assume it's got something to do with pointers, but I'd like to know more since it's used in so many places in Godot's codebase).

As for which version I prefer I think the one with incrementation in the loop seems a bit more clear than the one with indexing.

Reply:
`tetrahedrons` seems to be a `Vector` which in turn uses copy on write "COW". The COW paradigm is used extensively in Godot source.

`Vector` in Godot is _not_ the same as e.g. an STL vector. For that we (confusingly) have `LocalVector`.

The COW implementation is rather heavyweight especially for writing individual elements, so for large numbers of elements or bottlenecks it is recommended to get a pointer `ptrw()` (write pointer) as a one off and use that to avoid COW machinery.

Reply:
Indeed, two parts here:
* `push_back` will cause multiple reallocations
* The structure is COW so any mutations cause some threading things, even if it's not shared

Reply:
Cheers, thanks for your thorough explanation @lawnjelly and nice perf optimization catch @AThousandShips, changes should be up and running in a sec :)

Reply:
Last note, given your explanation we might want to revisit the 2D implementation and apply the same optimizations over there:

https://github.com/godotengine/godot/blob/master/core/math/geometry_2d.h#L305-L315